src/crimson/osd/exceptions.h

```diff
@@ -60,3 +60,525 @@ struct input_output_error : public error {
+  static auto handle(Func&& func) {
+    return [
+      func = std::forward<Func>(func)
+    ] (const unthrowable_wrapper&) mutable -> decltype(auto) {
+      return std::invoke(std::forward<Func>(func));
+    };
+  }
+
+private:
+  // can be used only to initialize the `instance` member
+  explicit unthrowable_wrapper() = default;
+
+  // implement the errorable interface
+  struct throwable_carrier{};
+  static std::exception_ptr carrier_instance;
+
+  static constexpr const std::type_info& _exception_ptr_type_info() {
+    return typeid(throwable_carrier);
+  }
+  auto _to_exception_ptr() const {
+    // error codes don't need to instantiate `std::exception_ptr` each
+    // time as the code is actually a part of the type itself.
+    // `std::make_exception_ptr()` on modern enough GCCs is quite cheap
+    // (see the Gleb Natapov's patch eradicating throw/catch there),
+    // but using one instance per type boils down the overhead to just
+    // ref-counting.
+    return carrier_instance;
+  }
+  static const auto& _from_exception_ptr(std::exception_ptr) {
+    return instance;
+  }
+
+  friend class error_t<unthrowable_wrapper<ErrorT, ErrorV>>;
+};
+
+template <class ErrorT, ErrorT ErrorV>
+std::exception_ptr unthrowable_wrapper<ErrorT, ErrorV>::carrier_instance = \
+  std::make_exception_ptr<
+    unthrowable_wrapper<ErrorT, ErrorV>::throwable_carrier>({});
+
+namespace _impl {
+  template <class T> struct always_false : std::false_type {};
+};
+
+template <class ErrorVisitorT, class FuturatorT>
+class maybe_handle_error_t {
+  const std::type_info& type_info;
+  typename FuturatorT::type result;
+  ErrorVisitorT errfunc;
+
+public:
+  maybe_handle_error_t(ErrorVisitorT&& errfunc, std::exception_ptr ep)
+    : type_info(*ep.__cxa_exception_type()),
+      result(FuturatorT::make_exception_future(std::move(ep))),
+      errfunc(std::forward<ErrorVisitorT>(errfunc)) {
+  }
+
+  template <class ErrorT>
+  void operator()() {
+    static_assert(std::is_invocable<ErrorVisitorT, ErrorT>::value,
+                  "provided Error Visitor is not exhaustive");
+    // Throwing an exception isn't the sole way to signalize an error
+    // with it. This approach nicely fits cold, infrequent issues but
+    // when applied to a hot one (like ENOENT on write path), it will
+    // likely hurt performance.
+    // Alternative approach for hot errors is to create exception_ptr
+    // on our own and place it in the future via make_exception_future.
+    // When ::handle_exception is called, handler would inspect stored
+    // exception whether it's hot-or-cold before rethrowing it.
+    // The main advantage is both types flow through very similar path
+    // based on future::handle_exception.
+    //
+    // NOTE: this is extension of language. It should be available both
+    // in GCC and Clang but a fallback (basing on throw-catch) could be
+    // added as well.
+    if (type_info == ErrorT::get_exception_ptr_type_info()) {
```

Comment:
With the usage below in safe_then, I'm a bit worried that duplicate types in WrappedAllowedErrorsT could cause the same (possibly stateful) handler to be called twice.  Is that possible?

Reply:
Yes, that's possible. I was thinking about two possibilities: deduplication or re-arranging the `maybe_handle_error_t` / `safe_then` to end processing on first match.

Reply:
Deduplication generally seems necessary since in other places you can't guarrantee that different safe_then arms have disjoint error returns.

Reply:
This also precludes the possibility of using error hierarchies, I think?  I'm probably ok with that, but it's probably worth calling out somewhere prominent that you always have to enumerate the concrete leaf types.

Reply:
Yes, that's correct. Maybe we could extract base classes (and bases of bases of...) at compile-time to call the check also for them but I doubt the performance trade-off and complexity burden are worth the feature.

More appealing approach could be to exclude/static assert such hierarchies.

Reply:
Hmm, another way is to differentiate: `__cxa_exception_type()` for trivial / final types, try + catch for hierarchical. Might be worth consideration if most of our hot-path types will be simple.

Reply:
I'm fine with this for now.  Even better if we could statically prevent it.